packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Snippets/BinaryDataSnippets.cs

```diff
@@ -0,0 +1,37 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Microsoft.Generator.CSharp.Expressions;
+using Microsoft.Generator.CSharp.Primitives;
+using static Microsoft.Generator.CSharp.Snippets.Snippet;
+
+namespace Microsoft.Generator.CSharp.Snippets
+{
+    public static class BinaryDataSnippets
+    {
+        public static ScopedApi ToObjectFromJson(this ScopedApi<BinaryData> binaryData, Type responseType)
+            => binaryData.Invoke(nameof(BinaryData.ToObjectFromJson), [], [new CSharpType(responseType)], false).As(responseType);
+
+        public static ListSnippet ToArray(this ScopedApi<BinaryData> binaryData) => new(typeof(byte[]), binaryData.Invoke(nameof(BinaryData.ToArray)));
+
+        public static ScopedApi<BinaryData> FromStream(ValueExpression stream, bool async)
+        {
+            var methodName = async ? nameof(BinaryData.FromStreamAsync) : nameof(BinaryData.FromStream);
+            return Static<BinaryData>().Invoke(methodName, [stream], null, callAsAsync: async).As<BinaryData>();
+        }
+
+        public static ValueExpression ToMemory(this ScopedApi<BinaryData> binaryData) => binaryData.Invoke(nameof(BinaryData.ToMemory));
```

Comment:
OK I see the trick here. We made all these "convenience methods bonding to a snippet with a specific type" as extension methods to make it easier to extend.
This is a genius idea. This should also solve the issue my previous PR trying to solve: https://github.com/microsoft/typespec/pull/3223 therefore I closed this PR, I think since last time we changed everything to be in the "fluent style", [the issue](https://github.com/microsoft/typespec/issues/3312) does not exist either.
Please feel free to close that issue as well